src/txorphanage.cpp

```diff
@@ -153,26 +159,43 @@ void TxOrphanage::AddChildrenToWorkSet(const CTransaction& tx, std::set<uint256>
 
 bool TxOrphanage::HaveTx(const GenTxid& gtxid) const
 {
-    LOCK(g_cs_orphans);
+    LOCK(m_mutex);
+
     if (gtxid.IsWtxid()) {
         return m_wtxid_to_orphan_it.count(gtxid.GetHash());
     } else {
         return m_orphans.count(gtxid.GetHash());
     }
 }
 
-std::pair<CTransactionRef, NodeId> TxOrphanage::GetTx(const uint256& txid) const
+bool TxOrphanage::GetTxToReconsider(NodeId peer, CTransactionRef& ref, NodeId& originator, bool& more)
 {
-    AssertLockHeld(g_cs_orphans);
-
-    const auto it = m_orphans.find(txid);
-    if (it == m_orphans.end()) return {nullptr, -1};
-    return {it->second.tx, it->second.fromPeer};
+    LOCK(m_mutex);
+
+    auto work_set_it = m_peer_work_set.find(peer);
+    if (work_set_it != m_peer_work_set.end()) {
+        auto& work_set = work_set_it->second;
+        auto it = work_set.begin();
+        while (it != work_set.end()) {
+            uint256 txid = *it;
+            it = work_set.erase(it);
+
+            const auto orphan_it = m_orphans.find(txid);
+            if (orphan_it != m_orphans.end()) {
+                more = (it != work_set.end());
```

Comment:
(if you're keeping `more`), the following is equivalent, and maybe more direct/clear:

```suggestion
                more = !(work_set.empty());
```